src/Control/Effect/Cut.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE DeriveFunctor, ExistentialQuantification, FlexibleContexts, FlexibleInstances, LambdaCase, MultiParamTypeClasses, StandaloneDeriving, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Cut
+( Cut(..)
+, cutfail
+, call
+, cut
+, Branch(..)
+, branch
+, runCut
+, CutC(..)
+) where
+
+import Control.Applicative (Alternative(..))
+import Control.Effect.Carrier
+import Control.Effect.Internal
+import Control.Effect.NonDet
+import Control.Effect.Sum
+
+-- | 'Cut' effects are used with 'NonDet' to provide control over backtracking.
+data Cut m k
+  = Cutfail
+  | forall a . Call (m a) (a -> k)
+
+deriving instance Functor (Cut m)
+
+instance HFunctor Cut where
+  hmap _ Cutfail    = Cutfail
+  hmap f (Call m k) = Call (f m) k
+
+instance Effect Cut where
+  handle _     _       Cutfail    = Cutfail
+  handle state handler (Call m k) = Call (handler (m <$ state)) (handler . fmap k)
+
+-- | Fail the current branch, and prevent backtracking within the nearest enclosing 'call' (if any).
+--
+--   Contrast with 'empty', which fails the current branch but allows backtracking.
+--
+--   prop> run (runNonDetOnce (runCut (cutfail <|> pure a))) == Nothing
+--   prop> run (runNonDetOnce (runCut (pure a <|> cutfail))) == Just a
+cutfail :: (Carrier sig m, Member Cut sig) => m a
+cutfail = send Cutfail
+
+-- | Delimit the effect of 'cutfail's, allowing backtracking to resume.
+--
+--   prop> run (runNonDetOnce (runCut (call (cutfail <|> pure a) <|> pure b))) == Just b
```

Comment:
Likewise, this test for `call`’s interaction with `cutfail` is pleasantly simple.